src/Symfony/Component/Security/Csrf/Tests/DoubleSubmitCsrfTokenManagerTest.php

```diff
@@ -0,0 +1,232 @@
+        $session->expects($this->once())->method('get')->with('x-csrf-token')->willReturn(2);
+        $this->logger->expects($this->once())->method('warning')->with('CSRF validation failed: double-submit info was used in a previous request but is now missing.');
+        $this->assertFalse($this->csrfTokenManager->isTokenValid($token));
+    }
+
+    public function testMissingPreviousOrigin()
+    {
+        $session = $this->createMock(Session::class);
+        $request = new Request();
+        $request->cookies->set('x-csrf-token_'.str_repeat('a', 24), 'x-csrf-token');
+        $request->setSession($session);
+        $request->cookies->set('sess', 'id');
+        $this->requestStack->push($request);
+
+        $token = new CsrfToken('test_token', str_repeat('a', 24));
+
+        $session->expects($this->once())->method('getName')->willReturn('sess');
+        $session->expects($this->once())->method('get')->with('x-csrf-token')->willReturn(1);
+        $this->logger->expects($this->once())->method('warning')->with('CSRF validation failed: origin info was used in a previous request but is now missing.');
+        $this->assertFalse($this->csrfTokenManager->isTokenValid($token));
+    }
+
+    public function testValidDoubleSubmit()
+    {
+        $request = new Request();
+        $request->cookies->set('x-csrf-token_'.str_repeat('a', 24), 'x-csrf-token');
+        $this->requestStack->push($request);
+
+        $token = new CsrfToken('test_token', str_repeat('a', 24));
+
+        $this->logger->expects($this->once())->method('debug')->with('CSRF validation accepted using double-submit info.');
+        $this->assertTrue($this->csrfTokenManager->isTokenValid($token));
+        $this->assertSame(2, $request->attributes->get('x-csrf-token'));
+    }
+
+    public function testCheckOnlyHeader()
+    {
+        $csrfTokenManager = new DoubleSubmitCsrfTokenManager($this->requestStack, $this->logger, DoubleSubmitCsrfTokenManager::CHECK_ONLY_HEADER);
+
+        $request = new Request();
+        $tokenValue = str_repeat('a', 24);
+        $request->headers->set('x-csrf-token', $tokenValue);
+        $this->requestStack->push($request);
+
+        $token = new CsrfToken('test_token', $tokenValue);
+
+        $this->logger->expects($this->once())->method('debug')->with('CSRF validation accepted using double-submit info.');
+        $this->assertTrue($csrfTokenManager->isTokenValid($token));
+        $this->assertSame('x-csrf-token', $request->cookies->get('x-csrf-token_'.$tokenValue));
+
+        $this->logger->expects($this->once())->method('warning')->with('CSRF validation failed: wrong token found in header info.');
+        $this->assertFalse($csrfTokenManager->isTokenValid(new CsrfToken('test_token', str_repeat('b', 24))));
+    }
+
+    /**
+     * @testWith [0]
+     *           [1]
+     *           [2]
+     */
+    public function testMissingDoubleSubmit(int $checkHeader)
```

Comment:
**note:** Reformulating this behaviour in plain English.

Without session, a valid `Origin` or `Referer` header is enough to accept the request.
It apply for all modes.

It is not possible, with configuration, to add a strict constraint on the header or cookie validity.
Or I missed something.

Reply:
Renamed to testValidOriginMissingDoubleSubmit

> It is not possible, with configuration, to add a strict constraint on the header or cookie validity

That's correct, we don't need this flexibility as this would just result in less people being able to interact with the app.

Reply:
> less people being able to interact with the app

Let's suppose that
- a customer have a supported browser.
- the behaviour of the app works as expected, no bugs, no misconfiguration, no network issue.

What issue can occur that avoid interacting with the app?

Reply:
Flexibility is complexity. It'd prefer waiting for a real world use case before making this configurable.

Reply:
First, we will need to migrate to Symfony 7 to use it.

This limitation is a blocker to adopt this double-submit protection.

Or just try backporting this CSRF token manager.